server/athenian/api/align/queries/goals.py

```diff
@@ -81,18 +70,73 @@ async def resolve_goals(
         cache=info.context.cache,
         slack=info.context.app["slack"],
     )
-    res = []
-    for i, req in enumerate(team_metrics_requests):
-        initial_metrics = all_metric_values[req.time_intervals[0]][req.metrics[0]]
-        current_metrics = all_metric_values[req.time_intervals[1]][req.metrics[0]]
+    return [to_serve.build_goal_tree(all_metric_values).to_dict() for to_serve in goals_to_serve]
+
+
+class _GoalToServe:
+    """A goal served in the response."""
+
+    def __init__(
+        self,
+        team_goal_rows: Sequence[Row],
+        team_tree: TeamTree,
+        team_member_map: dict[int, list[int]],
+        only_with_targets: bool,
+    ):
+        self._team_goal_rows = team_goal_rows
+        self._request, self._goal_team_tree = self._parse_rows(
+            team_goal_rows, team_tree, team_member_map, only_with_targets,
+        )
+
+    @property
+    def request(self) -> TeamMetricsRequest:
+        return self._request
+
+    def build_goal_tree(self, metric_values) -> GoalTree:
+        initial_metrics = metric_values[self._request.time_intervals[0]][self._request.metrics[0]]
+        current_metrics = metric_values[self._request.time_intervals[1]][self._request.metrics[0]]
         metric_values = GoalMetricValues(initial_metrics, current_metrics)
-        team_goal_rows = team_goal_rows_per_request[i]
-        group_res = _team_tree_to_goal_tree(
-            team_tree, team_goal_rows[0], team_goal_rows, metric_values,
+        return _team_tree_to_goal_tree(
+            self._goal_team_tree, self._team_goal_rows[0], self._team_goal_rows, metric_values,
         )
-        res.append(group_res)
 
-    return [r.to_dict() for r in res]
+    @classmethod
+    def _parse_rows(
+        cls,
+        team_goal_rows: Sequence[Row],
+        team_tree: TeamTree,
+        team_member_map: dict[int, list[int]],
+        only_with_targets: bool,
+    ) -> tuple[TeamMetricsRequest, TeamTree]:
+        goal_row = team_goal_rows[0]  # could be any, all rows have the joined Goal columns
+        metric = TEMPLATES_COLLECTION[goal_row[Goal.template_id.name]]["metric"]
+
+        valid_from = goal_row[Goal.valid_from.name]
+        expires_at = goal_row[Goal.expires_at.name]
+        initial_interval = goal_initial_query_interval(valid_from, expires_at)
+
+        if only_with_targets:
+            # if the flag is true
+            # 1 - prune team tree to keep only branches with assigned teams
+            keep_team = _TeamFilterFromTeamGoalRows(team_goal_rows)
+            goal_team_tree = _team_tree_prune_empty_branches(team_tree, keep_team)
+            # assigned_teams is a non empty subset of all the ids in team_tree,
+            # so the pruned tree is never empty
+            assert goal_team_tree is not None
+            # 2 - in the metrics requests only ask for teams present in the pruned tree
+            goal_tree_team_ids = goal_team_tree.get_flatten_team_ids()
```

Comment:
```suggestion
            goal_tree_team_ids = goal_team_tree.flatten_team_ids()
```

or

```suggestion
            goal_tree_team_ids = goal_team_tree.get_flat_team_ids()
```

I like the first variant more.